File: projects/m5/003-probability-calculator/solution/PisapiaDario/prob_calculator.py
import copy
import random

class Hat:
    def __init__(self, **kwargs):
        self.kwargs = kwargs
        self.contents = []
        for key, value in self.kwargs.items():
            for value in range(value):
                self.contents.append(key)

    def draw(self, balls):                                          
        drawn = []
        if balls < len(self.contents):
            for n in range(balls):
                x = random.choice(self.contents)
                self.contents.remove(x)
                drawn.append(x)
            return drawn
        else:
            return self.contents
  
            
def experiment(hat, expected_balls, num_balls_drawn, num_experiments):
    balls_cont = []
    for key, value in expected_balls.items():
        for value in range(value):
            balls_cont.append(key)

    y = 0        
    for n in range(num_experiments):
        hat_copy = copy.deepcopy(hat)
        balls_check = copy.copy(balls_cont)
        balls_drawn = hat_copy.draw(num_balls_drawn)
        for ball in balls_drawn:
            if ball in balls_check:
                balls_check.remove(ball)
        if balls_check == []:
            y += 1  
       
    return y/num_experiments     

File: projects/m5/003-probability-calculator/solution/valentinaBazzacchi/prob-calculator.js
const lodashClonedeep = require("lodash.clonedeep")

class Hat {
    constructor(...args) {
        this.contents = [];

        args.forEach((arg) => {
            for (const [key, value] of Object.entries(arg)) {
                for (let i = 0; i < value; i++) {
                    this.contents.push(key);
                }
            }
        })

    } // end constructor

    draw(numOfBalls) {
        
        if (numOfBalls > this.contents.length) return this.contents;

        const ballsDrawed = [];
        for (let i = 0; i < numOfBalls; i++) {
            const choice = randomChoice(this.contents);
            ballsDrawed.push(choice);
            const startIndex = this.contents.indexOf(choice);
            this.contents.splice(startIndex, 1);
        }
        return ballsDrawed;
    }
} // end class

/* Each experiment consists of:

* Start with a hat containing the specified balls
* The hat object should be copied before each draw
* Draw a number of balls ("numBallsDrawn") with the "draw" method
* Check if we got the balls we were attempting to draw ("expectedBalls") and count the successes
* The "experiment" function should return a probability (successes / numer of experiments)

*/

function experiment(hat, expectedBalls, numBallsDrawn, numExperiments) {
    
    let successes = 0;
    
    for (let i = 0; i < numExperiments; i++) {
        // const hatCopy = JSON.parse(JSON.stringify(hat)); --> doesn't work for class methods copy
        const hatCopy = lodashClonedeep(hat);
        const ballsDrawed = hatCopy.draw(numBallsDrawn);
        
        let success = true;
        for (const [key, value] of Object.entries(expectedBalls)) {
            if (count(ballsDrawed, key) < value) success = false;
        }
        
        successes += success ? 1 : 0;
    } // end for loop
    
    return successes / numExperiments;
} // end function


// ------------------------  UTILS -------------------------------

// Function to simulate the draw of a random ball from the hat
function randomChoice(list) {
    const randomIndex = Math.floor((Math.random() * list.length));
    return list[randomIndex];
}

// Function to count the number of equal elements contained in a list
function count(arr, item) {
    return arr.filter(value => value === item).length;
}

module.exports = { Hat, experiment }